### ISSUE TITLE
duplicate selected features

### ISSUE THREAD

**Issue by Songweiping** [jsvir/lscae#1]
Hi,

Thanks for the cool work! When I run this code on my own data, I find that the selected features are not unique. In my case, I have around 1600 features, and I try to select a subset of features using your algorithm. However, there are 880 unique features if I set the k_selected paramter to 1500. The duplication is true even if i choose a small subset of features, for example, 200 out of 1600. 

My question is, how can I get exact K distinct features if k_selected==K? I want to figure out the minimal number of features needed for my task.

**Comment by jsvir** [comment]
Hi,

Have you applied standard scaler on your dataset before the train?
In addition, can you share a subset of your data so we could reproduce your results?

Best regards

**Comment by Songweiping** [comment]
Thanks for your reply.

I use standard scaler on my dataset before training. 

Sorry I cannot share my data for some reason, but the phenomenon can be reproduced using the two moon data. I modify your orginal notebook in following ways: 1) set the parameter d in create_twomoon_dataset to 1600 (the number of features in my data), 2) set num_epochs to 30 (for fast reproduce). 3) select 1500 features instead of 2. Finally, I only get 1047 unique features although the algorithm outputs 1500 features. 

I think the problem could be in the selection procedure,
`        feats = torch.argmax(self.logits, dim=0)`
In my mind, this can not guarantee the uniqueness of selected features as the largest number of different columns may appear in the same row. Maybe I miss something here.

Another interesting phenomenon is, when I set d to 100 in the two moon dataset, even if I want to select two features, I didn't get the 0 and 1 indexes (which I should have).

**Comment by jsvir** [comment]
Thx so much for pointing out this issue. We have updated the code, accordingly.
Please note that the code does not guarantee that the number of unique features will be exactly k_selected, however the gap between  the two will be much smaller than before.
Also, args.fr is a hyperparameter penalizing feature redundancy.
We recommend setting its value to 1e1, and train your model for 500 epochs.

Regarding the selected features in the two_moons experiments - we are aware of the fact that the two original features are not always selected when the number of nuisance features exceeds 30 or so.
